{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This code was developed and authored by Jerzy Twarowski in Malkova Lab at the University of Iowa \n",
    "# Contact: jerzymateusz-twarowski@uiowa.edu, tvarovski1@gmail.com\n",
    "\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import seaborn as sns\n",
    "import umap\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import cancer_config as cfg\n",
    "import logging\n",
    "\n",
    "#parameters#################\n",
    "\n",
    "cancer = cfg.settings[\"TCGA-PROJECT\"]\n",
    "pickle_path = f\"expression_data_{cancer}.pickle\"\n",
    "df_consolidated_path = f\"consolidated_results_{cancer}.tsv\"\n",
    "username = cfg.settings[\"username\"]\n",
    "\n",
    "metadata_file = f\"TCGA-{cancer}-WXS-BAM-metadata.tsv\"\n",
    "metadata_location = f\"/Users/{username}/MMBIR_Databases/TCGA/{metadata_file}\"\n",
    "\n",
    "df_diff_expr_path = f\"outputs/ttest_results_{cancer}_minconc0_bh_corrected.tsv\"\n",
    "\n",
    "sample_type = \"Primary Tumor\"\n",
    "count = \"Raw_Count\"\n",
    "n_neighbors = 50\n",
    "min_dist = 0.1\n",
    "n_components = 2\n",
    "metric = \"canberra\"\n",
    "\n",
    "filter_transcriptome = True\n",
    "min_pval = 0.05\n",
    "fold_change_lower = 0.5\n",
    "fold_change_upper = 2\n",
    "\n",
    "scale_upper = 700\n",
    "scale_lower = 50\n",
    "\n",
    "#############################\n",
    "\n",
    "def filterByDiffExpr(df, df_diff_expr, min_pval=0.1, fold_change_lower=0.75, fold_change_upper=1.5):\n",
    "\n",
    "    logging.basicConfig(level=logging.INFO)\n",
    "\n",
    "    # take only the genes where the p-value is less than min_pval\n",
    "    df_diff_expr = df_diff_expr[df_diff_expr[\"p-value\"] < min_pval]\n",
    "    logging.info(f\"There are now {df_diff_expr.shape[0]} differentially expressed transcripts with p-value < {min_pval}.\")\n",
    "\n",
    "    # take only the genes where the fold change is outside fold_change_lower and fold_change_upper\n",
    "    df_diff_expr = df_diff_expr[(df_diff_expr[\"fold-change\"] < fold_change_lower) | (df_diff_expr[\"fold-change\"] > fold_change_upper)]\n",
    "    logging.info(f\"There are now {df_diff_expr.shape[0]} differentially expressed transcripts with fold-change < {fold_change_lower} or > {fold_change_upper}.\")\n",
    "\n",
    "    #get the gene_id of the differentially expressed genes\n",
    "    diff_expr_genes = df_diff_expr[\"gene_id\"].tolist()\n",
    "\n",
    "    #add columns to keep to the list\n",
    "    diff_expr_genes.extend(['sample_name', 'sample_name_file', 'case_id'])\n",
    "\n",
    "    #drop the columns from the df that are not in the list\n",
    "    df = df[diff_expr_genes]\n",
    "\n",
    "    return df\n",
    "\n",
    "def addMMBToDF(df, df_consolidated, sample_type, count):\n",
    "\n",
    "    logging.basicConfig(level=logging.INFO)\n",
    "\n",
    "    # filter the df_consolidated dataframe to only include primary tumors\n",
    "    df_consolidated = df_consolidated[df_consolidated[\"Sample_Type\"] == sample_type]\n",
    "\n",
    "    #group df_consolidated by Case_ID and get the highest value of Raw_Count and Filtered_Count\n",
    "    grouping_dict = {\"Raw_Count\": \"max\", \"Filtered_Count\": \"max\"}\n",
    "    df_consolidated = df_consolidated.groupby(\"Case_ID\").agg(grouping_dict).reset_index()\n",
    "\n",
    "    #add a column with Raw_Count to the df by merging on the Case_ID column (in df_consolidated) and the case_id column (in df)\n",
    "    df = df.merge(df_consolidated[[\"Case_ID\", count]], left_on=\"case_id\", right_on=\"Case_ID\")\n",
    "    logging.info(f\"Added the {count} column to the dataframe.\")\n",
    "\n",
    "    #remove the Case_ID column\n",
    "    df = df.drop(columns=[\"Case_ID\"])\n",
    "\n",
    "    return df\n",
    "\n",
    "def addMetadataToDF(df, df_metadata, columns):\n",
    "    \n",
    "    logging.basicConfig(level=logging.INFO)\n",
    "\n",
    "    df = df.merge(df_metadata[[\"cases.0.case_id\"]+columns], left_on=\"case_id\", right_on=\"cases.0.case_id\")\n",
    "    logging.info(f\"Added the {columns} columns to the dataframe.\")\n",
    "\n",
    "    #remove the \"cases.0.case_id\" column\n",
    "    df = df.drop(columns=[\"cases.0.case_id\"])\n",
    "    \n",
    "    return df\n",
    "\n",
    "def perform_UMAP(rnaseq_data, n_neighbors, min_dist, n_components, metric):\n",
    "\n",
    "    logging.basicConfig(level=logging.INFO)\n",
    "\n",
    "    # first, create a umap object\n",
    "    reducer = umap.UMAP(        \n",
    "            n_neighbors=n_neighbors,\n",
    "            min_dist=min_dist,\n",
    "            n_components=n_components,\n",
    "            metric=metric)\n",
    "\n",
    "    logging.info(\"Scaling the data\")\n",
    "    scaled_rnaseq_data = StandardScaler().fit_transform(rnaseq_data)\n",
    "    logging.info(\"Done scaling the data\")\n",
    "\n",
    "    # then, fit the umap object to the data\n",
    "    logging.info(\"Fitting the umap object to the data\")\n",
    "    embedding = reducer.fit_transform(scaled_rnaseq_data)\n",
    "    logging.info(\"Done fitting the umap object to the data\")\n",
    "    logging.info(f\"The shape of the embedding is {embedding.shape}\")\n",
    "    logging.info(\"Done performing UMAP, returning the embedding\")\n",
    "\n",
    "    return reducer, embedding\n",
    "\n",
    "def plot_UMAP_numerical(embedding, df, numerical_col, scale_upper=700, scale_lower=0, save_path=None, show_plot=True):\n",
    "\n",
    "    logging.basicConfig(level=logging.INFO)\n",
    "\n",
    "    # plot the umap projection, coloring by the Raw_Count column, set the highest value of color to 1000\n",
    "    plt.figure(figsize=(12,12))\n",
    "    # assign marker by figo_stage\n",
    "    plt.scatter(embedding[:, 0], embedding[:, 1], c=df[numerical_col], s=16, alpha=0.8, cmap='viridis', vmin=scale_lower, vmax=scale_upper)\n",
    "    plt.title(f'UMAP of RNAseq by {count}', fontsize=24, fontweight='bold')\n",
    "    plt.xlabel(\"UMAP 1\", fontsize=18)\n",
    "    plt.ylabel(\"UMAP 2\", fontsize=18)\n",
    "    plt.axis('square')\n",
    "    plt.colorbar(label=f\"{count}\")\n",
    "    #save the figure\n",
    "    if save_path:\n",
    "        plt.savefig(f\"outputs/umap_{cancer}_by_{count}.png\", dpi=600)\n",
    "        logging.info(f\"Saved the figure to {save_path}\")\n",
    "    \n",
    "    if show_plot:\n",
    "        plt.show()\n",
    "    else:\n",
    "        plt.close()\n",
    "\n",
    "def plot_UMAP_categorical(embedding, df, categorical_col, save_path=None, show_plot=True):\n",
    "\n",
    "    logging.basicConfig(level=logging.INFO)\n",
    "\n",
    "    color_labels = df[categorical_col].unique() # get the unique values of the column\n",
    "    rgb_values = sns.color_palette(\"tab10\", len(color_labels)) # create a list of rgb values\n",
    "    color_map = dict(zip(color_labels, rgb_values)) # create a dictionary of the color labels and the rgb values\n",
    "    logging.info(f\"The color labels are {color_labels}\")\n",
    "\n",
    "    plt.figure(figsize=(12,12))\n",
    "    # assign color by column_var which is a string, so it will be categorical\n",
    "    plt.scatter(embedding[:, 0], embedding[:, 1], c=df[categorical_col].map(color_map), s=16, alpha=0.5)\n",
    "    plt.axis('square')\n",
    "\n",
    "    plt.title(f'UMAP of RNAseq by {categorical_col}', fontsize=24)\n",
    "    plt.xlabel(\"UMAP 1\", fontsize=18)\n",
    "    plt.ylabel(\"UMAP 2\", fontsize=18)\n",
    "\n",
    "    # reuse the color_map for the legend\n",
    "    patches = [mpatches.Patch(color=color, label=label) for label, color in color_map.items()]\n",
    "    plt.legend(handles=patches, bbox_to_anchor=(1, 1), loc=2, borderaxespad=0.)\n",
    "    plt.tight_layout()\n",
    "\n",
    "    if save_path:\n",
    "        plt.savefig(f\"outputs/umap_{cancer}_by_{column_var}.png\", dpi=600)\n",
    "        logging.info(f\"Saved the figure to {save_path}\")\n",
    "    \n",
    "    if show_plot:\n",
    "        plt.show()\n",
    "    else:\n",
    "        plt.close()\n",
    "\n",
    "\n",
    "# load the data from a pickle file, the consolidated results file, and the differentially expressed genes file\n",
    "df = pd.read_pickle(pickle_path)\n",
    "df_consolidated = pd.read_csv(df_consolidated_path, sep=\"\\t\")\n",
    "df_diff_expr = pd.read_csv(df_diff_expr_path, sep=\"\\t\")\n",
    "df_metadata = pd.read_csv(metadata_location, sep=\"\\t\")\n",
    "\n",
    "trailing_columns = 4 # the number of columns that are not part of the data, but are added by the script or are metadata\n",
    "\n",
    "if filter_transcriptome:\n",
    "\n",
    "    df = filterByDiffExpr(df, df_diff_expr, min_pval=min_pval, fold_change_lower=fold_change_lower, fold_change_upper=fold_change_upper)\n",
    "\n",
    "df = addMMBToDF(df, df_consolidated, sample_type, count)\n",
    "\n",
    "# add the specified metadata to the dataframe\n",
    "column_var_list_numerical = []\n",
    "column_var_list_categorical = [\"cases.0.disease_type\", \"cases.0.diagnoses.0.figo_stage\", \"cases.0.samples.0.portions.0.analytes.0.aliquots.0.concentration\"]\n",
    "#\"cases.0.samples.0.portions.0.analytes.0.aliquots.0.submitter_id\"\n",
    "\n",
    "column_var_list=column_var_list_numerical + column_var_list_categorical\n",
    "\n",
    "df = addMetadataToDF(df, df_metadata, column_var_list)\n",
    "trailing_columns += len(column_var_list)\n",
    "\n",
    "print(df.columns)\n",
    "print(f\"There are {len(df.columns)} columns in the dataframe and {trailing_columns} are trailing\")\n",
    "\n",
    "\n",
    "# create a list of the columns that we want to use for the umap projection\n",
    "#it is all columns except the last few columns which are not part of the data\n",
    "rnaseq_data = df.iloc[:, :-trailing_columns]\n",
    "\n",
    "reducer, embedding = perform_UMAP(rnaseq_data, n_neighbors, min_dist, n_components, metric)\n",
    "\n",
    "plot_UMAP_numerical(embedding, \n",
    "                    df, \n",
    "                    count, \n",
    "                    scale_upper=scale_upper, \n",
    "                    scale_lower=scale_lower, \n",
    "                    save_path=f\"outputs/umap_{cancer}_by_{count}.png\", \n",
    "                    show_plot=True)\n",
    "\n",
    "for column_var in column_var_list_categorical:\n",
    "\n",
    "    if column_var == \"cases.0.diagnoses.0.figo_stage\":\n",
    "\n",
    "        #replace the column_var names, specifically remove A, B, C, 1, 2\n",
    "        df[column_var] = df[column_var].str.replace(\"A\", \"\")\n",
    "        df[column_var] = df[column_var].str.replace(\"B\", \"\")\n",
    "        df[column_var] = df[column_var].str.replace(\"C\", \"\")\n",
    "        df[column_var] = df[column_var].str.replace(\"1\", \"\")\n",
    "        df[column_var] = df[column_var].str.replace(\"2\", \"\")\n",
    "\n",
    "    plot_UMAP_categorical(embedding,\n",
    "                        df, \n",
    "                        column_var, \n",
    "                        save_path=f\"outputs/umap_{cancer}_by_{column_var}.png\", \n",
    "                        show_plot=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "name": "python",
   "version": "3.10.4"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "b9ff68f6d6a3b24e01b1e0ddf31dd6e945ba7d66b325e9634af996f4d302e8f1"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
